protocols/mdns/src/behaviour/socket.rs

```diff
@@ -0,0 +1,149 @@
+        fn poll_receive_packet(
+            &mut self,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>> {
+            // Poll receive socket.
+            let _ = futures::ready!(self.poll_readable(cx));
+            match self.recv_from(buf).now_or_never() {
+                Some(Ok((len, from))) => Poll::Ready(Ok(Some((len, from)))),
+                Some(Err(err)) => Poll::Ready(Err(err)),
+                None => Poll::Ready(Ok(None)),
+            }
+        }
+
+        /// Attempts to send data on the socket to a given address.
+        fn poll_send_packet(
+            &mut self,
+            cx: &mut Context,
+            packet: &[u8],
+            to: SocketAddr,
+        ) -> Poll<Result<(), Error>> {
+            let _ = futures::ready!(self.poll_writable(cx));
+            match self.send_to(packet, to).now_or_never() {
+                Some(Ok(_)) => Poll::Ready(Ok(())),
+                Some(Err(err)) => Poll::Ready(Err(err)),
+                None => Poll::Pending,
+            }
+        }
+    }
+}
+
+#[cfg(feature = "tokio")]
+pub mod tokio {
+    use super::*;
+    use ::tokio::net::UdpSocket as TkUdpSocket;
+
+    /// Tokio ASync Socket`
+    pub type TokioUdpSocket = TkUdpSocket;
+
+    impl AsyncSocket for TokioUdpSocket {
+        /// Create the async socket from the ```std::net::UdpSocket```
+        fn from_socket(socket: UdpSocket) -> std::io::Result<Self> {
+            socket.set_nonblocking(true)?;
+            TokioUdpSocket::from_std(socket)
+        }
+
+        /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
+        fn poll_receive_packet(
+            &mut self,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>> {
+            match self.poll_recv_ready(cx) {
+                Poll::Ready(Ok(_)) => match self.try_recv_from(buf) {
+                    Ok((len, from)) => Poll::Ready(Ok(Some((len, from)))),
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        Poll::Ready(Ok(None))
+                    }
+                    Err(err) => Poll::Ready(Err(err)),
+                },
+                Poll::Ready(Err(e)) => Poll::Ready(Err(e)),
+                _ => Poll::Pending,
+            }
+        }
+
+        /// Attempts to send data on the socket to a given address.
+        fn poll_send_packet(
+            &mut self,
+            cx: &mut Context,
+            packet: &[u8],
+            to: SocketAddr,
+        ) -> Poll<Result<(), Error>> {
+            match self.poll_send_ready(cx) {
+                Poll::Ready(Ok(_)) => match self.try_send_to(packet, to) {
+                    Ok(_len) => Poll::Ready(Ok(())),
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => Poll::Ready(Ok(())),
```

Comment:
The [example](https://docs.rs/tokio/1.20.0/tokio/net/struct.UdpSocket.html#example-6) uses a loop here to avoid the `WouldBlock` case. I think we should do that too.